File: finished/unique_paths.js
// https://leetcode.com/problems/unique-paths/
//
/**
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
var uniquePaths = function (m, n) {
  return nChooseR(m + n - 2, m - 1);
};

function nChooseR(n, r) {
  if (r <= 0 || r >= n) return 1;
  return factorial(n) / (factorial(r) * factorial(n - r));
}

function factorial(n) {
  if (n <= 1) return n;

  let result = 1;

  while (n >= 1) {
    result *= n;
    n--;
  }

  return result;
}

File: src/python/finished/unique_paths.py
# https://leetcode.com/problems/unique-paths


class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        if m < n:
            m, n = n, m

        # m >= n
        tmp = [0] * n
        dp = [0] * n
        dp[0] = 1

        for _ in range(m):
            for c in range(n):
                tmp[c] = dp[c]
                if c > 0:
                    tmp[c] += tmp[c - 1]
            dp, tmp = tmp, dp
        return dp[-1]
